doc/designs/quic-design/quic-io-arch.md

```diff
@@ -0,0 +1,485 @@
+      allows an application to provide its own implementation of a
+      select(3)-like function. In fact, this has the potential to be quite
+      powerful and would allow the application to implement its own pollable
+      BIOs, and therefore perform blocking I/O on top of any custom BIO.
+
+      For example, while historically none of our own memory-based BIOs have
+      supported blocking semantics, a sophisticated application could if it
+      wished choose to implement a custom blocking memory BIO and implement a
+      custom poller which synchronises using a custom poll descriptor based
+      around condition variables rather than sockets. Thus this scheme is
+      highly flexible.
+
+      (It is worth noting also that the implementation of blocking semantics at
+      the application level also does not rely on any privileged access to the
+      internals of the QUIC implementation and an application could if it wished
+      build blocking semantics out of a non-blocking QUIC instance; this is not
+      particularly difficult, though providing custom pollers here would mean
+      there should be no need for an application to do so.)
+
+  - Configuring a socket into non-blocking mode might confuse an application.
+
+    However:
+
+    - Applications will already have to make changes to any network-side BIOs,
+      for example switching from a `BIO_s_socket` to a `BIO_s_dgram`, or from a
+      BIO pair to a `BIO_s_dgram_pair`. Custom BIOs will need to be
+      substantially reworked to switch from bytestream semantics to datagram
+      semantics. Such applications will already need substantial changes, and
+      this is unavoidable.
+
+      Of course, application impacts and migration guidance can (and will) all
+      be documented.
+
+    - In order for an application to be confused by us putting a socket into
+      non-blocking mode, it would need to be trying to use the socket in some
+      way. But it is not possible for an application to pass a socket to our
+      QUIC implementation, and also try to use the socket directly, and have
+      QUIC still work. Using QUIC necessarily requires that an application not
+      also be trying to make use of the same socket.
+
+    - There are some circumstances where an application might want to multiplex
+      other protocols onto the same UDP socket, for example with protocols like
+      RTP/RTCP or STUN; this can be facilitated using the QUIC fixed bit.
+      However, these use cases cannot be supported without explicit assistance
+      from a QUIC implementation and this use case cannot be facilitated by
+      simply sharing a network socket, as incoming datagrams will not be routed
+      correctly. (We may offer some functionality in future to allow this to be
+      coordinated but this is not for MVP.) Thus this also is not a concern.
+      Moreover, it is extremely unlikely that any such applications are using
+      sockets in blocking mode anyway.
+
```

Comment:
Some discussion of application controlled event loop with blocking sockets is needed somewhere. The consequence of requiring non-blocking sockets is that this won't be possible. I guess the argument is that this doesn't really matter because if the application has an event loop anyway it shouldn't really matter whether the underlying socking is blocking or non-blocking.

Reply:
What does "application controlled event loop with blocking sockets" even mean? If the sockets are blocking by definition the application isn't in control of its event loop.

If there is a public application which is a good example of this, it would be helpful.

Reply:
It seems like the reasonable way to handle this would be to provide an `SSL_stream_select` call (which, as mentioned in the thread below, we will need to provide to allow multiplexing on multiple streams eventually anyway).

I am not really sure how we could make this work as is with QUIC (i.e., if `s_client` continues to use `select(3)`) because just because the UDP network socket is readable, it doesn't mean the stream is. We would need to provide our own select API.